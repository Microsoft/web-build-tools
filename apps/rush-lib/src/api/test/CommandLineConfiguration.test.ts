// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { CommandLineConfiguration } from '../CommandLineConfiguration';

describe('CommandLineConfiguration', () => {
  it('Forbids a misnamed phase', () => {
    expect(
      () =>
        new CommandLineConfiguration({
          phases: [
            {
              name: '_faze:A',
              summary: 'A'
            }
          ]
        })
    ).toThrowErrorMatchingSnapshot();
    expect(
      () =>
        new CommandLineConfiguration({
          phases: [
            {
              name: '_phase:',
              summary: 'A'
            }
          ]
        })
    ).toThrowErrorMatchingSnapshot();
  });

  it('Detects a missing phase', () => {
    expect(
      () =>
        new CommandLineConfiguration({
          commands: [
            {
              commandKind: 'phased',
              name: 'example',
              summary: 'example',
              description: 'example',
              safeForSimultaneousRushProcesses: false,

              phases: ['_phase:A']
            }
          ]
        })
    ).toThrowErrorMatchingSnapshot();
  });

  it('Detects a missing phase dependency', () => {
    expect(
      () =>
        new CommandLineConfiguration({
          phases: [
            {
              name: '_phase:A',
              summary: 'A',
              dependencies: {
                upstream: ['_phase:B']
              }
            }
          ]
        })
    ).toThrowErrorMatchingSnapshot();

    expect(
      () =>
        new CommandLineConfiguration({
          phases: [
            {
              name: '_phase:A',
              summary: 'A',
              dependencies: {
                self: ['_phase:B']
              }
            }
          ]
        })
    ).toThrowErrorMatchingSnapshot();
  });

  it('Detects a cycle among phases', () => {
    expect(
      () =>
        new CommandLineConfiguration({
          phases: [
            {
              name: '_phase:A',
              summary: 'A',
              dependencies: {
                self: ['_phase:B']
              }
            },
            {
              name: '_phase:B',
              summary: 'C',
              dependencies: {
                self: ['_phase:C']
              }
            },
            {
              name: '_phase:C',
              summary: 'C',
              dependencies: {
                self: ['_phase:A']
              }
            }
          ]
        })
    ).toThrowErrorMatchingSnapshot();

    expect(
      () =>
        new CommandLineConfiguration({
          phases: [
            {
              name: '_phase:A',
              summary: 'A',
              dependencies: {
                upstream: ['_phase:B']
              }
            },
            {
              name: '_phase:B',
              summary: 'C',
              dependencies: {
                upstream: ['_phase:C']
              }
            },
            {
              name: '_phase:C',
              summary: 'C',
              dependencies: {
                upstream: ['_phase:A']
              }
            }
          ]
        })
    ).toThrowErrorMatchingSnapshot();
  });
});
