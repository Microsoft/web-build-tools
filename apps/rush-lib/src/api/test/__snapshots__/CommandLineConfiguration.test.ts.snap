// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandLineConfiguration Detects a cycle among phases 1`] = `"In command-line.json, there exists a cycle within the set of _phase:B dependencies: _phase:B, _phase:C, _phase:A"`;

exports[`CommandLineConfiguration Detects a cycle among phases 2`] = `"In command-line.json, there exists a cycle within the set of _phase:B dependencies: _phase:B, _phase:C, _phase:A"`;

exports[`CommandLineConfiguration Detects a missing phase 1`] = `"In command-line.json, in the command \\"example\\", the phase \\"_phase:A\\" does not exist."`;

exports[`CommandLineConfiguration Detects a missing phase dependency 1`] = `"In command-line.json, in the phase \\"_phase:A\\", the upstream dependency phase \\"_phase:B\\" does not exist."`;

exports[`CommandLineConfiguration Detects a missing phase dependency 2`] = `"In command-line.json, in the phase \\"_phase:A\\", the self dependency phase \\"_phase:B\\" does not exist."`;

exports[`CommandLineConfiguration Forbids a misnamed phase 1`] = `"In command-line.json, the phase \\"_faze:A\\"'s name does not begin with the required prefix \\"_phase:\\"."`;

exports[`CommandLineConfiguration Forbids a misnamed phase 2`] = `"In command-line.json, the phase \\"_phase:\\"'s name must have characters after \\"_phase:\\""`;
